### PR TITLE
Update to OpenCensus version 0.22.3

### PR DESCRIPTION
**Pull request by aLekSer**
This should fix Prometheus error messages on exporting with empty tags
list . Allocator Metrics.

Takeover this PR  #893
Part of #1330 

Fixing this error in logs (happened periodically):
```
textPayload: "2020/02/07 15:14:14 Failed to export to Prometheus: inconsistent label cardinality: expected 1 label values but got 0 in []string(nil)
```

### CODE REVIEW

**Comment by aLekSer** [comment]
Tested this change with Grafana and Prometheus.
Error with `inconsistent label cardinality` now does not happen in the logs.
https://github.com/googleforgames/agones/issues/1330#issuecomment-584737095

**Comment by markmandel** [comment]
@roberthbailey @pooneh-m should we aim to merge this in this release, or hold it until next release to give it time to bake? (Since RC is next Tuesday)

I'm inclined to push this to the next release, just to give it time to be tested over time. WDYT?

**Comment by aLekSer** [comment]
Setting it to on-hold. Test on Stackdriver failed:
```
2020/04/03 16:11:56 Failed to export to Stackdriver: rpc error: code = InvalidArgument desc = Field timeSeries[17].points[0].distributionValue had an invalid value: Distribution |explicit_buckets.bounds| does not have at least one entry.
```

**Comment by aLekSer** [comment]
Strange enough, but I was fixing similar thing a while ago:
https://github.com/googleforgames/agones/pull/554

**Comment by roberthbailey** [comment]
@aLekSer - can you split the vendor changes into a separate commit from the changes to pkg/ to make it easier to see how much impact there is on our code?

@markmandel - at first glance this seems to be making non-trivial code changes to our metric exporting which makes me tend to agree to push it to the next release to give it time to bake. On the other hand, how much extra testing does that actually get us? how much testing of this code path happens during our e2es? and would anyone test a pre-release build to verify the functionality on top of what the author and reviewer do? If not and it isn't destabilizing, then that would argue to get it in now since we would only get feedback on how well it works once it's released.

**Comment by aLekSer** [comment]
Similar error is mentioned here also:
https://github.com/census-ecosystem/opencensus-go-exporter-stackdriver/issues/234

**Comment by markmandel** [comment]
Just taking a look - at first glance LGTM, @cyriltovena do you have time to double check this, since it was originally your code?

**Comment by cyriltovena** [comment]
I'll run through it quickly now

File: pkg/gameserverallocations/metrics.go [hunk 1]
```diff
@@ -54,12 +54,12 @@ func init() {
 
 // default set of tags for latency metric
 var latencyTags = []tag.Mutator{
-	tag.Insert(keyMultiCluster, "none"),
-	tag.Insert(keyClusterName, "none"),
-	tag.Insert(keySchedulingStrategy, "none"),
-	tag.Insert(keyFleetName, "none"),
-	tag.Insert(keyNodeName, "none"),
-	tag.Insert(keyStatus, "none"),
+	tag.Insert(keyMultiCluster, ""),
```
Comment (cyriltovena) [inline pkg/gameserverallocations/metrics.go:57]: I'm guessing now this is allowed to be empty ?

Comment (aLekSer) [reply pkg/gameserverallocations/metrics.go:57]: Insert() should add this tag https://github.com/census-instrumentation/opencensus-go/blob/master/tag/map.go#L107:6
I will check tomorrow on how this applies to this metrics. Probably add a test for this.

**Comment by cyriltovena** [comment]
image
image
image
image
image
image
image
image
image
image

**Comment by cyriltovena** [comment]
The only one that does not work is the go-client cache. But I assume I did not need cache for this try out ?

**Comment by aLekSer** [comment]
Thanks @cyriltovena so much for providing screenshots, I need to fix stackdriver error message which is left for this PR.
```
2020/04/03 16:11:56 Failed to export to Stackdriver: rpc error: code = InvalidArgument desc = Field timeSeries[17].points[0].distributionValue had an invalid value: Distribution |explicit_buckets.bounds| does not have at least one entry.
```

File: pkg/gameserverallocations/metrics.go (hunk 1 above)
Comment (aLekSer) [reply pkg/gameserverallocations/metrics.go]: @cyriltovena This change break Stackdriver exporter, so I readded `"none"` for now. I think this change could be done separately.

**Comment by aLekSer** [comment]
With "none" in tag values export to Stackdriver is working.
Stackdriver output

**Comment by aLekSer** [comment]
Debugged opencensus stackdriver exporter code a bit, added more logging:
```
E 2020-04-13T07:28:37.571709353Z 2020/04/13 07:28:37 Failed to export this to Stackdriver 0: gameservers_count
E 2020-04-13T07:28:37.571744133Z 2020/04/13 07:28:37 Failed to export this to Stackdriver 1: k8s_client_workqueue_unfinished_work_seconds
E 2020-04-13T07:28:37.571749105Z 2020/04/13 07:28:37 Failed to export this to Stackdriver 2: fleets_replicas_count
E 2020-04-13T07:28:37.571752888Z 2020/04/13 07:28:37 Failed to export this to Stackdriver 3: k8s_client_workqueue_work_duration_seconds
E 2020-04-13T07:28:37.571796974Z 2020/04/13 07:28:37 Failed to export to Stackdriver: rpc error: code = InvalidArgument desc = Field timeSeries[8].points[0].distributionValue had an invalid value: Distribution |explicit_buckets.bounds| does not have at least one entry.
```
Stackdriver export is not working on the most recent stackdriver exporter also.

The error with Stackdriver metrics is gone only when I comment out whole file: https://github.com/googleforgames/agones/blob/master/pkg/metrics/kubernetes_client.go
Need to understand which metric or group of them lead to this situation, because error on exporter like `timeSeries[8]` does not help to find erroneous metric.

**Comment by aLekSer** [comment]
Narrowing down - `k8s.io/client-go/util/workqueue` metrics are failing with Stackdriver exporter `workqueue.SetProvider(c)`.

**Comment by aLekSer** [comment]
Was able to fix Stackdriver exporter uploading if parameter is changed from `view.Distribution(0)` or `view.Distribution()` (both cause an error like above) to something like:
```
	runtime.Must(view.Register(&view.View{
		Name:        "k8s_client_workqueue_latency_seconds",
		Measure:     workQueueLatencyStats,
		Description: "How long an item stays in the work queue.",
		Aggregation: view.Distribution(0.0001),
		TagKeys:     []tag.Key{keyQueueName},
	}))
```

**Comment by markmandel** [comment]
How is this PR looking? I think it's good to merge, but wanted to check first.

**Comment by aLekSer** [comment]
This PR will fix errors in Prometheus exporter, and I will produce a separate PR for fixing Stackdriver exporter `view.Distribution(..)` - with multiple parameters. (Even adding 0.0001 instead of 0 is fixing Stackdriver)

**Pull request merged**